crates/bevy_pbr/src/meshlet/cull_meshlets.wgsl

```diff
@@ -3,36 +3,61 @@
-/// 1. The first pass is only frustum culling, on only the clusters that were visible last frame.
-/// 2. The second pass performs both frustum and occlusion culling (using the depth buffer generated from the first pass), on all clusters.
+/// 1. The first pass tests instance visibility, frustum culling, LOD selection, and finally occlusion culling using last frame's depth pyramid.
+/// 2. The second pass performs occlusion culling (using the depth buffer generated from the first pass) on all clusters that passed
+///    the instance, frustum, and LOD tests in the first pass, but were not visible last frame according to the occlusion culling.
 
 @compute
 @workgroup_size(128, 1, 1) // 128 threads per workgroup, 1 instanced meshlet per thread
-fn cull_meshlets(@builtin(global_invocation_id) cluster_id: vec3<u32>) {
-    // Fetch the instance data and check for instance culling
-    if cluster_id.x >= arrayLength(&meshlet_thread_meshlet_ids) { return; }
-    let instance_id = meshlet_thread_instance_ids[cluster_id.x];
-    if should_cull_instance(instance_id) {
-        return;
-    }
+fn cull_meshlets(
+    @builtin(workgroup_id) workgroup_id: vec3<u32>,
+    @builtin(num_workgroups) num_workgroups: vec3<u32>,
+    @builtin(local_invocation_id) local_invocation_id: vec3<u32>,
+) {
+    // Calculate the cluster ID for this thread
+    let cluster_id = local_invocation_id.x + 128u * dot(workgroup_id, vec3(num_workgroups.x * num_workgroups.x, num_workgroups.x, 1u));
+    if cluster_id >= arrayLength(&meshlet_thread_meshlet_ids) { return; }
+
+#ifdef MESHLET_SECOND_CULLING_PASS
+    if !meshlet_is_second_pass_candidate(cluster_id) { return; }
+#endif
+
+    // Check for instance culling
+    let instance_id = meshlet_thread_instance_ids[cluster_id];
+#ifdef MESHLET_FIRST_CULLING_PASS
+    if should_cull_instance(instance_id) { return; }
+#endif
 
-    // Fetch other meshlet data
-    let meshlet_id = meshlet_thread_meshlet_ids[cluster_id.x];
+    // Calculate world-space culling bounding sphere for the cluster
     let instance_uniform = meshlet_instance_uniforms[instance_id];
+    let meshlet_id = meshlet_thread_meshlet_ids[cluster_id];
     let model = affine3_to_square(instance_uniform.model);
     let model_scale = max(length(model[0]), max(length(model[1]), length(model[2])));
     let bounding_spheres = meshlet_bounding_spheres[meshlet_id];
+    var culling_bounding_sphere_center = model * vec4(bounding_spheres.self_culling.center, 1.0);
+    var culling_bounding_sphere_radius = model_scale * bounding_spheres.self_culling.radius;
+
+#ifdef MESHLET_FIRST_CULLING_PASS
+    // Frustum culling
+    // TODO: Faster method from https://vkguide.dev/docs/gpudriven/compute_culling/#frustum-culling-function
+    for (var i = 0u; i < 6u; i++) {
+        if dot(view.frustum[i], culling_bounding_sphere_center) <= -culling_bounding_sphere_radius {
```

Comment:
For consistency with the CPU frustum sphere intersection code:
```suggestion
        if dot(view.frustum[i], culling_bounding_sphere_center) + culling_bounding_sphere_radius <= 0.0 {
```

Reply:
I'll have to check that this doesn't hurt performance. Performance is extremely important here, as it runs for 50+ million clusters a frame :). I've micro-optimized every shader quite a bit (and still have more to go, the bounding sphere calculations have too many live registers).